VariantShredding.md

```diff
@@ -25,276 +25,302 @@
+          optional double typed_value;
+        }
+      }
+    }
+    required group tags {
+      optional binary value;
+      optional group typed_value (LIST) {
+        repeated group list {
+          required group element {
+            optional binary value;
+            optional binary typed_value (STRING);
+          }
         }
-       }
       }
-     }
-     optional group b {
-      optional int64 typed_value;
-      optional binary variant_value;
-     }
     }
-   }
   }
- }
 }
 ```
 
-In the above example schema, if “a” is an array containing a mix of integer and non-integer values, the engine will shred individual elements appropriately into either `typed_value` or `variant_value`.
-If the top-level Variant is not an array (for example, an object), the engine cannot shred the value and it will store it in the top-level `variant_value`.
-Similarly, if "a" is not an array, it will be stored in the `variant_value` under "a".
+# Data Skipping
 
-Consider the following example:
+Statistics for `typed_value` columns can be used for file, row group, or page skipping when `value` is always null (missing).
 
-```
-[
-  {
-    "a": [1, 2, 3],
-    "b": 100,
-    “c”: “unexpected”
-  },
-  {
-    "a": [4, 5, 6],
-    "b": 200
-  },
-  “not an object”
-]
-```
+When the corresponding `value` column is all nulls, all values must be the shredded `typed_value` field's type.
+Because the type is known, comparisons with values of that type are valid.
+`IS NULL`/`IS NOT NULL` and `IS NAN`/`IS NOT NAN` filter results are also valid.
 
-The second array element can be fully shredded, but the first and third cannot be. The contents of `variant_col.array[*].variant_value` would be as follows:
+Comparisons with values of other types are not necessarily valid and data should not be skipped.
 
-```
-[
-  { “c”: “unexpected” },
-  NULL,
-  “not an object”
-]
+Casting behavior for Variant is delegated to processing engines.
+For example, the interpretation of a string as a timestamp may depend on the engine's SQL session time zone.
+
+## Reconstructing a Variant
+
+It is possible to recover a full Variant value using a recursive algorithm, where the initial call is to `construct_variant` with the top-level Variant group fields.
+
+```python
+def construct_variant(metadata, value, typed_value):
+    """Constructs a Variant from value and typed_value"""
+    if typed_value is not None:
+        if isinstance(typed_value, dict):
+            # this is a shredded object
+            object_fields = {
+                name: construct_variant(metadata, field.value, field.typed_value)
+                for (name, field) in typed_value
+            }
+
+            if value is not None:
+                # this is a partially shredded object
+                assert isinstance(value, VariantObject), "partially shredded value must be an object"
+                assert typed_value.keys().isdisjoint(value.keys()), "object keys must be disjoint"
```

Comment:
I know the rules above say it may return an error here or pull the value out of "typed_value". But if we are going to not allow it in this reference code I probably would say we should just never allow it